job/service.go

```diff
@@ -738,3 +692,194 @@ func populateDownstreamDAGs(dagTree *tree.MultiRootTree, jobSpec models.JobSpec,
 
 	return rootNode, nil
 }
+
+// Refresh fetches all the requested jobs, resolves its dependencies, assign proper priority weights,
+// compile all jobs in the project and upload them to the destination store.
+func (srv *Service) Refresh(ctx context.Context, projectSpec models.ProjectSpec, namespaceJobNamePairs []models.NamespaceJobNamePair,
+	progressObserver progress.Observer) (err error) {
+	// Resolve dependency
+	if err = srv.resolveAndPersistDependency(ctx, projectSpec, namespaceJobNamePairs, progressObserver); err != nil {
+		// if err is caused by dependency resolution, ignore this as error.
+		var merrs *multierror.Error
+		if !errors.As(err, &merrs) {
+			return err
+		}
+	}
+
+	// Fetch dependency and enrich
+	jobDependencies, err := srv.dependencyResolver.FetchJobDependencies(ctx, projectSpec, progressObserver)
+	if err != nil {
+		return fmt.Errorf("failed to fetch job dependencies: %w", err)
+	}
+
+	// Get all job specs and enrich with dependencies
+	projectJobSpecRepo := srv.projectJobSpecRepoFactory.New(projectSpec)
+	jobSpecs, err := srv.prepareAllJobsInProject(ctx, jobDependencies, projectJobSpecRepo)
+	if err != nil {
+		return err
+	}
+
+	// Resolve priority
+	jobSpecs, err = srv.priorityResolver.Resolve(ctx, jobSpecs, progressObserver)
+	if err != nil {
+		return err
+	}
+
+	// Group per namespace
+	namespaceJobSpecMap, err := prepareNamespaceJobsMap(ctx, projectJobSpecRepo, jobSpecs)
+	if err != nil {
+		return err
+	}
+
+	// Compile & Deploy
+	for namespace, namespaceJobSpecs := range namespaceJobSpecMap {
+		namespaceSpec, err := srv.namespaceService.Get(ctx, projectSpec.Name, namespace)
+		if err != nil {
+			return err
+		}
+
+		if err := srv.batchScheduler.DeployJobs(ctx, namespaceSpec, namespaceJobSpecs, progressObserver); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (srv *Service) resolveAndPersistDependency(ctx context.Context, projectSpec models.ProjectSpec,
+	namespaceJobNamePairs []models.NamespaceJobNamePair, progressObserver progress.Observer) (err error) {
+	start := time.Now()
+	defer resolveDependencyHistogram.Observe(time.Since(start).Seconds())
+
+	jobSpecs, err := srv.prepareJobSpecs(ctx, projectSpec, namespaceJobNamePairs)
+	if err != nil {
+		return err
+	}
+	srv.notifyProgress(progressObserver, &models.ProgressJobSpecFetch{})
+
+	// compile assets before resolving in parallel
+	for i, jSpec := range jobSpecs {
+		if jobSpecs[i].Assets, err = srv.assetCompiler(jSpec, srv.Now()); err != nil {
+			return fmt.Errorf("asset compilation: %w", err)
+		}
+	}
+
+	// resolve specs in parallel
+	runner := parallel.NewRunner(parallel.WithTicket(ConcurrentTicketPerSec), parallel.WithLimit(ConcurrentLimit))
+	for _, jobSpec := range jobSpecs {
```

Comment:
If possible we can see a mechanism on abstract out all the operations of how the parallel running is happening by keeping things declarative

Reply:
decided to not refactor this part in this PR, as abstracting this out is quite tricky, and will see a better way to do this.